flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/SqlITCaseBase.java

```diff
@@ -190,4 +231,76 @@ private static List<String> readResultFiles(Path path) throws IOException {
         }
         return result;
     }
+
+    protected List<String> formatRawResult(List<String> rawResults) {
+        return rawResults;
+    }
+
+    protected static List<String> convertToMaterializedResult(
+            List<String> rawResults,
+            ResolvedSchema schema,
+            DeserializationSchema<RowData> deserializationSchema) {
+        DataCollector collector = new DataCollector();
+        try {
+            deserializationSchema.open(new TestingDeserializationContext());
+            for (String rawResult : rawResults) {
+                deserializationSchema.deserialize(rawResult.getBytes(), collector);
+            }
+        } catch (Exception e) {
+            fail("deserialize error: ", e);
+        }
+
+        RowRowConverter converter = RowRowConverter.create(schema.toPhysicalRowDataType());
+        Map<Row, Row> upsertResult = new HashMap<>();
+
+        for (RowData rowData : collector.dataList) {
+            RowKind kind = rowData.getRowKind();
+
+            Row row = converter.toExternal(rowData);
+            assertThat(row).isNotNull();
+
+            Row key = Row.project(row, schema.getPrimaryKeyIndexes());
+            key.setKind(RowKind.INSERT);
+
+            Row upsertRow = Row.copy(row);
+            upsertRow.setKind(RowKind.INSERT);
```

Comment:
`upsertRow` is only useful when the kind is insert or update_after. So we could create it when `kind == RowKind.INSERT || kind == RowKind.UPDATE_AFTER`.